### ISSUE TITLE
INSTALL_PARSE_FAILED_NO_CERTIFICATES

### ISSUE THREAD

**Issue by jinlei-du** [calabash/calabash-android#578]
I am the new to calabash-android, and have no idea where it's from of following error.  My .calabash_settings also has been updated. 
 
Can TestServer.apk be installed to device directly, it's still failed even i tried to install it by hand.

{"package_name":"","activity_name":"","app_path":"","api_level":"","keystore_location":"C:\Users\dujinlei\.android\debug.keystore","keystore_password":"android","keystore_alias":"androiddebugkey","keystore_alias_password":"android"}
 

2015-05-28 16:24:17 - JDK found at: C:\Program Files\Java\jdk1.7.0_67
2015-05-28 16:24:17 - Android SDK found at: D:\tool\adt-bundle-windows-x86_64-20131030\sdk
2015-05-28 16:24:17 - Trying to find launchable activity
2015-05-28 16:24:17 - Found tools directory at 'D:/tool/adt-bundle-windows-x86_64-20131030/sdk/build-tools/android-4.4'
2015-05-28 16:24:17 - Found launchable activity 'sh.calaba.demoproject.AndroidDemoProjectActivity'
2015-05-28 16:24:17 - D:/tool/ruby-1.9.3-p551-i386-mingw32/bin/ruby.exe -S cucumber -v MAIN_ACTIVITY=sh.calaba.demoproject.AndroidDemoProjectActivity APP_PATH="D:/code/calabash-android-demoapp-master/calabash/AndroidDemoProject.apk" TEST_APP_PATH="test_servers/5712a70e81ad038459247d5e0a5c7b81_0.5.9.apk"
Code:
  * features/support/env.rb
  * features/support/app_installation_hooks.rb
  * features/support/app_life_cycle_hooks.rb
  * features/support/hooks.rb
  * features/step_definitions/calabash_steps.rb

Features:
  * features/my_first.feature
Parsing feature files took 0m0.009s

Feature: Demo feature

  Scenario: I can start my app # features\my_first.feature:3
2015-05-28 16:24:18 - First scenario in feature - reinstalling apps
2015-05-28 16:24:18 - connected_devices: ["4dfa00fc2502317b"]
2015-05-28 16:24:18 - "D:\tool\adt-bundle-windows-x86_64-20131030\sdk/platform-tools/adb.exe" -s 4dfa00fc2502317b forward tcp:34777 tcp:7102
2015-05-28 16:24:18 - 
2015-05-28 16:24:18 - Found tools directory at 'D:/tool/adt-bundle-windows-x86_64-20131030/sdk/build-tools/android-4.4'
2015-05-28 16:24:18 - Uninstalling: sh.calaba.demoproject.test
2015-05-28 16:24:18 - tttttt
2015-05-28 16:24:19 - Failure

2015-05-28 16:24:20 - Found tools directory at 'D:/tool/adt-bundle-windows-x86_64-20131030/sdk/build-tools/android-4.4'
2015-05-28 16:24:21 - Uninstalling: sh.calaba.demoproject
2015-05-28 16:24:21 - tttttt
2015-05-28 16:24:22 - Failure

2015-05-28 16:24:23 - test_servers/5712a70e81ad038459247d5e0a5c7b81_0.5.9.apk
2015-05-28 16:24:23 - Installing: test_servers/5712a70e81ad038459247d5e0a5c7b81_0.5.9.apk
2015-05-28 16:24:23 - Installing: test_servers/5712a70e81ad038459247d5e0a5c7b81_0.5.9.apk
2015-05-28 16:24:24 - 	pkg: /data/local/tmp/5712a70e81ad038459247d5e0a5c7b81_0.5.9.apk

Failure [INSTALL_PARSE_FAILED_NO_CERTIFICATES]

2015-05-28 16:24:24 - Found tools directory at 'D:/tool/adt-bundle-windows-x86_64-20131030/sdk/build-tools/android-4.4'
  undefined method `last' for #<Enumerator:0x2db5b18> (NoMethodError)
  D:/tool/ruby-1.9.3-p551-i386-mingw32/lib/ruby/gems/1.9.1/gems/calabash-android-0.5.9/lib/calabash-android/operations.rb:316:in `install_app'
  D:/tool/ruby-1.9.3-p551-i386-mingw32/lib/ruby/gems/1.9.1/gems/calabash-android-0.5.9/lib/calabash-android/operations.rb:105:in `install_app'
  D:/code/calabash-android-demoapp-master/calabash/features/support/app_installation_hooks.rb:23:in `Before'
  unexpected return (LocalJumpError)
  D:/code/calabash-android-demoapp-master/calabash/features/support/app_life_cycle_hooks.rb:11:in `Before'
2015-05-28 16:24:26 - It looks like your app is no longer running. 
It could be because of a crash or because your test script shut it down.
2015-05-28 16:24:26 - Server not responding. Moving on.

Failing Scenarios:
cucumber features\my_first.feature:3 # Scenario: I can start my app

1 scenario (1 failed)
0 steps
0m7.772s

**Comment by jinlei-du** [comment]
re-install 0.5.8, it works.  Maybe it's the bug of 0.5.9

**Comment by TobiasRoikjer** [comment]
@jinlei-du Are you resigning your apk using Calabash?

**Comment by istvanbb** [comment]
I was just about to report this problem, I have the same issue. It's probably a bug in **calabash-android 0.5.9** and it's happening for Android **4.2.2** and **4.1.1** devices. (other device versions are ok)

To reproduce:
1. Build you apk file. You can *adb install* the apk to a device with no problem.
2. Do a *calabash-android resign* on the apk. If you try to *adb install* the newly signed apk to a 4.2.2 or 4.1.1 device, you will get INSTALL_PARSE_FAILED_NO_CERTIFICATES error.

Possible workarounds:
If you have 0.5.8 installed you can force gem version by doing:
*calabash-android \_0.5.8\_ resign your-path-to-apk*

You can use 0.5.9 for the actual execution of tests as normal.

**Comment by TobiasRoikjer** [comment]
@istvanbb What signing algorithm does your keystore use?

**Comment by istvanbb** [comment]
@TobiasRoikjer  SHA256withRSA

**Comment by TobiasRoikjer** [comment]
@istvanbb We have noticed that some Android devices do not seem to support SHA256withRSA. Could you verify that the apk you have (before Calabash has touched it at all) is signed with SHA256withRSA, and is able to be installed on the 4.2.2 and 4.1.1 devices?

**Comment by istvanbb** [comment]
@TobiasRoikjer Yes, I checked the pre-calabash apk with
*unzip -p Name-of-apk.apk META-INF/CERT.RSA | keytool -printcert*

and it tells me it's SHA256withRSA. It can be installed without any problems.

**Comment by jinlei-du** [comment]
@TobiasRoikjer Yes, I have resigned apk

**Comment by jinlei-du** [comment]
@TobiasRoikjer  Another small problem i would like to raise up is that following command to extract information may invalid on non-English OS(I see this problem in windows), since JDK tool prompts according OS region settings. It can add option "-J"-Duser.language=en to work around .

[calabash-android/ruby-gem/lib/calabash-android/java_keystore.rb]

line 15:        keystore_data = system_with_stdout_on_success(Env.keytool_path, '-list', '-v', '-keystore', location, '-storepass', password, '-J"-Dfile.encoding=utf-8"')

D:\code\calabash-android-demoapp-master\calabash> "C:\Program Files\Java\jdk1.7.
0_67/bin/keytool.exe" -list -v -alias androiddebugkey -keystore C:/Users/dujinle
i/.android/debug.keystore -storepass android "-J"-Dfile.encoding=utf-8""
别名: androiddebugkeygkey
创建日期: 2013-10-3013-10-30
条目类型: PrivateKeyEntryKeyEntry
证书链长度: 1�长度: 1�: 1
证书[1]:[1]:
所有者: CN=Android Debug, O=Android, C=US
发布者: CN=Android Debug, O=Android, C=US

**Comment by TobiasRoikjer** [comment]
@istvanbb I have sent you an email.

**Comment by TobiasRoikjer** [comment]
@jinlei-du Using English locale is a great suggestion. I think we will do so.

**Comment by TobiasRoikjer** [comment]
@jinlei-du Does upgrading to 0.5.10.pre1 fix the locale issue?

**Comment by TobiasRoikjer** [comment]
@jinlei-du  0.5.10.pre2, sorry, not 0.5.10.pre1

**Comment by thaitujim** [comment]
still use calabash-android 0.5.8. it's working fine !

**Comment by istvanbb** [comment]
@TobiasRoikjer I got your email. I will try to build an example app to demonstrate the bug for you guys. I was sick and didn't have the time.

**Comment by jinlei-du** [comment]
@TobiasRoikjer sorry for replying late, hope it's still useful to u. I tried latest version v0.5.10 ,  locale issue was fixed

**Comment by TobiasRoikjer** [comment]
@istvanbb This should be fixed in 0.5.11. From the changelog:
```
Change the way Calabash-Android (re)signs applications:

 - Use the encryption of the supplied keystore, and always use SHA1 as the digest algorithm. This is equivalent to how Android Studio behaves. 

    If given a keystore with the signing algorithm SHA256withDSA, Calabash-Android will sign an application using SHA1withDSA.
```
